src/main/java/org/jenkinsci/main/modules/sshd/CLICommandAdapter.java

```diff
@@ -21,22 +23,44 @@
     public Command create(CommandLine commandLine) {
         String cmd = commandLine.get(0);
         final CLICommand c = CLICommand.clone(cmd);
-        if (c==null)        return null;    // no such command
+        if (c == null) {
+            return null;    // no such command
+        }
 
         return new AsynchronousCommand(commandLine) {
+            private ExitCallback callback;
+
+            public void setExitCallback(ExitCallback callback) {
+                this.callback = callback;
+            }
+
             @Override
-            protected int run() throws Exception {
+            public void start(ChannelSession channel, Environment env) throws IOException {
                 // run as the authenticated user if we've actually authenticated the user,
                 // or otherwise run as anonymous
                 User u = getCurrentUser();
-                if (u!=null)    c.setTransportAuth(u.impersonate());
+                if (u!=null){
+                    c.setTransportAuth(u.impersonate());
+                }
 
                 CommandLine cmds = getCmdLine();
 
                 //TODO: Consider switching to UTF-8
-                return c.main(cmds.subList(1,cmds.size()), Locale.getDefault(), getInputStream(),
-                        new PrintStream(getOutputStream(), false, Charset.defaultCharset().toString()),
-                        new PrintStream(getErrorStream(), false, Charset.defaultCharset().toString()));
+                int ret = c.main(cmds.subList(1,cmds.size()), Locale.getDefault(), getInputStream(),
+                              new PrintStream(getOutputStream(), false, Charset.defaultCharset().toString()),
+                              new PrintStream(getErrorStream(), false, Charset.defaultCharset().toString()));
+                callback.onExit(ret);
+                channel.close();
+            }
+
+            @Override
+            public void destroy(ChannelSession channel) throws Exception {
+            }
+
+            //TODO run is no longer needed
+            @Override
+            protected int run() throws Exception {
+                return 0;
```

Comment:
so
```suggestion
                throw new IllegalStateException("no longer used");
```
?

Reply:
I vote to remove it from org.jenkinsci.main.modules.sshd.AsynchronousCommand

Reply:
I have reviewed this again, and the current implementation requires changes on other plugins to work. Digging a little why the channel is closed if you spawn a new thread and the start method ends, I have found that the [CommandLifecycle recommends implementing Runnable](https://github.com/apache/mina-sshd/blob/master/sshd-core/src/main/java/org/apache/sshd/server/command/CommandLifecycle.java#L32-L33) , [checking some implementations](https://github.com/apache/mina-sshd/blob/master/sshd-core/src/main/java/org/apache/sshd/server/shell/InvertedShellWrapper.java#L174-L221) it is also possible to implement a method that waits for the end of the command by checking the thread, I do not like too much because we will have a thread waiting for another thread doing nothing, but this implementation does not require changes in other plugins. 
In the case of implement runnable the signature of the class changes because already it has a method `protected int run()` that does not match with the runnable version `public void run()`. On the other hand, because of the command change of package, it is needed to make PRs to other plugins to change an import, so I am not sure if make a simple change (wait for the thread) or a good change (implement runnable).
I will make another PR with the runnable implementation and in this, I will implement the thread wait. I will give you the decision.

@jglick @timja

Reply:
if we're making a breaking change why not go for the cleaner implementation (the other PR)?

Reply:
I agree, I want to know @jglick opinion in case I miss something